src/testharness/TestHarness.cpp

```diff
@@ -4,27 +4,77 @@
 #include "tests/TestRunning.h"
 #include "util.h"
 
+#include <chrono>
 #include <filesystem>
 #include <iostream>
+#include <optional>
 #include <sstream>
+#include <thread>
 #include <utility>
 
 namespace tester {
 
+void swap(TestResult& first, TestResult& second) {
+  std::swap(first, second);
+}
+
 // Builds TestSet during object creation.
 bool TestHarness::runTests() {
+  // initialize the threads
+  std::thread t(&TestHarness::spawnThreads, this);
+
   bool failed = false;
   // Iterate over executables.
   for (auto exePair : cfg.getExecutables()) {
     // Iterate over toolchains.
     for (auto& tcPair : cfg.getToolChains()) {
-      if (runTestsForToolChain(exePair.first, tcPair.first) == 1)
+      if (aggregateTestResultsForToolChain(tcPair.first, exePair.first) == 1)
         failed = true;
     }
   }
+
+  // join the control thread
+  t.join();
   return failed;
 }
 
+void TestHarness::spawnThreads() {
+  int16_t numThreads = 0;
+  std::vector<std::thread> threadPool;
+
+  // Initialize the threads
+  // Iterate over executables.
+  for (auto exePair : cfg.getExecutables()) {
+    // Iterate over toolchains.
+    for (auto& tcPair : cfg.getToolChains()) {
+      // iterate over packages
+      for (auto& package : testSet) { // TestSet.second -> Package
+        // iterate over subpackages
+        for (auto& subpackage : package.second) { // Package.second -> SubPackage
+          // If we have already spawned the maximum number of threads,
+          //  wait for the first one to finish before spawning another.
+          if (numThreads >= cfg.getNumThreads()) {
+            threadPool.back().join();
+            threadPool.pop_back();
+            numThreads--;
+          }
+
+          // spawn a new thread executing its tests
+          std::thread t(&TestHarness::threadRunTestsForToolChain, this, tcPair.first, exePair.first, std::ref(subpackage.second));
```

Comment:
You can just use `subpackage.second` if you convert the parameter to `SubPackage&`.